File: cc/subtle/nonce_based_streaming_aead.cc
// Copyright 2019 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/subtle/nonce_based_streaming_aead.h"

#include <utility>

#include "absl/strings/string_view.h"
#include "tink/input_stream.h"
#include "tink/output_stream.h"
#include "tink/random_access_stream.h"
#include "tink/streaming_aead.h"
#include "tink/subtle/decrypting_random_access_stream.h"
#include "tink/subtle/stream_segment_decrypter.h"
#include "tink/subtle/stream_segment_encrypter.h"
#include "tink/subtle/streaming_aead_decrypting_stream.h"
#include "tink/subtle/streaming_aead_encrypting_stream.h"
#include "tink/util/statusor.h"

namespace crypto {
namespace tink {
namespace subtle {

crypto::tink::util::StatusOr<std::unique_ptr<crypto::tink::OutputStream>>
    NonceBasedStreamingAead::NewEncryptingStream(
        std::unique_ptr<crypto::tink::OutputStream> ciphertext_destination,
        absl::string_view associated_data) const {
  auto segment_encrypter_result = NewSegmentEncrypter(associated_data);
  if (!segment_encrypter_result.ok()) return segment_encrypter_result.status();
  return StreamingAeadEncryptingStream::New(
      std::move(segment_encrypter_result.ValueOrDie()),
      std::move(ciphertext_destination));
}

crypto::tink::util::StatusOr<std::unique_ptr<crypto::tink::InputStream>>
    NonceBasedStreamingAead::NewDecryptingStream(
        std::unique_ptr<crypto::tink::InputStream> ciphertext_source,
        absl::string_view associated_data) const {
  auto segment_decrypter_result = NewSegmentDecrypter(associated_data);
  if (!segment_decrypter_result.ok()) return segment_decrypter_result.status();
  return StreamingAeadDecryptingStream::New(
      std::move(segment_decrypter_result.ValueOrDie()),
      std::move(ciphertext_source));
}

crypto::tink::util::StatusOr<std::unique_ptr<crypto::tink::RandomAccessStream>>
    NonceBasedStreamingAead::NewDecryptingRandomAccessStream(
        std::unique_ptr<crypto::tink::RandomAccessStream> ciphertext_source,
        absl::string_view associated_data) const {
  auto segment_decrypter_result = NewSegmentDecrypter(associated_data);
  if (!segment_decrypter_result.ok()) return segment_decrypter_result.status();
  return DecryptingRandomAccessStream::New(
      std::move(segment_decrypter_result.ValueOrDie()),
      std::move(ciphertext_source));
}

}  // namespace subtle
}  // namespace tink
}  // namespace crypto

File: java_src/src/main/java/com/google/crypto/tink/subtle/NonceBasedStreamingAead.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.subtle;

import com.google.crypto.tink.StreamingAead;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.ReadableByteChannel;
import java.nio.channels.SeekableByteChannel;
import java.nio.channels.WritableByteChannel;
import java.security.GeneralSecurityException;

/**
 * An abstract class for StreamingAead using the nonce based online encryption scheme proposed in <a
 * href="https://eprint.iacr.org/2015/189.pdf">Online Authenticated-Encryption and its Nonce-Reuse
 * Misuse-Resistance</a> by Hoang, Reyhanitabar, Rogaway and Vizár.
 */
abstract class NonceBasedStreamingAead implements StreamingAead {

  // Abstract methods that the subclass has to implement.
  public abstract StreamSegmentEncrypter newStreamSegmentEncrypter(byte[] associatedData)
      throws GeneralSecurityException;
  // TODO(bleichen): Consider to pass aad here too.
  public abstract StreamSegmentDecrypter newStreamSegmentDecrypter()
      throws GeneralSecurityException;

  public abstract int getPlaintextSegmentSize();

  public abstract int getCiphertextSegmentSize();

  public abstract int getCiphertextOffset();

  public abstract int getCiphertextOverhead();

  public abstract int getHeaderLength();

  @Override
  public WritableByteChannel newEncryptingChannel(
      WritableByteChannel ciphertextChannel, byte[] associatedData)
      throws GeneralSecurityException, IOException {
    return new StreamingAeadEncryptingChannel(this, ciphertextChannel, associatedData);
  }

  @Override
  public ReadableByteChannel newDecryptingChannel(
      ReadableByteChannel ciphertextChannel, byte[] associatedData)
      throws GeneralSecurityException, IOException {
    return new StreamingAeadDecryptingChannel(this, ciphertextChannel, associatedData);
  }

  @Override
  public SeekableByteChannel newSeekableDecryptingChannel(
      SeekableByteChannel ciphertextSource, byte[] associatedData)
      throws GeneralSecurityException, IOException {
    return new StreamingAeadSeekableDecryptingChannel(this, ciphertextSource, associatedData);
  }

  @Override
  public OutputStream newEncryptingStream(OutputStream ciphertext, byte[] associatedData)
      throws GeneralSecurityException, IOException {
    return new StreamingAeadEncryptingStream(this, ciphertext, associatedData);
  }

  @Override
  public InputStream newDecryptingStream(InputStream ciphertextStream, byte[] associatedData)
      throws GeneralSecurityException, IOException {
    return new StreamingAeadDecryptingStream(this, ciphertextStream, associatedData);
  }
}
